src/Tools/dotnet-dump/Program.cs

```diff
@@ -1,61 +1,49 @@
+                         .AddCommand(CollectCommand())
+                         .AddCommand(AnalyzeCommand())
+                         .UseVersionOption()
+                         .UseHelp()
+                         .UseParseDirective()
+                         .UseDebugDirective()
+                         .UseSuggestDirective()
+                         .RegisterWithDotnetSuggest()
+                         .UseParseErrorReporting()
+                         .UseExceptionHandler()
+                         .Build();
+
+            return parser.InvokeAsync(args);
+        }
 
-            // Get the process
-            var process = Process.GetProcessById(ProcessId);
+        private static Command CollectCommand() =>
+            new Command("collect", "Captures memory dumps of .NET processes.", new Option[] { ProcessIdOption(), OutputOption() }, handler: CommandHandler.Create<IConsole, int, string>(new Dumper().Collect));
 
-            // Generate the file name
-            var fileName = Path.Combine(OutputDir, $"{process.ProcessName}-{process.Id}-{DateTime.Now:yyyyMMdd-HHmmss-fff}.dmp");
+        private static Option ProcessIdOption() =>
+            new Option(new[] { "-p", "--process-id" }, "The ID of the process to collect a memory dump.", new Argument<int> { Name = "processId" });
 
-            console.WriteLine($"Collecting memory dump for {process.ProcessName} (ID: {process.Id}) ...");
-            await Dumper.CollectDumpAsync(process, fileName);
-            console.WriteLine($"Dump saved to {fileName}");
+        private static Option OutputOption() =>
+            new Option(new[] { "-o", "--output" }, "The directory to write the dump. Defaults to the current working directory.", new Argument<string>(Directory.GetCurrentDirectory()) { Name = "directory" });
 
-            return 0;
-        }
+        private static Command AnalyzeCommand() =>
+            new Command("analyze", "Start interactive dump analyze.", new Option[] { RunCommand() }, argument: DumpFile(), handler: CommandHandler.Create<FileInfo, string[]>(new Analyzer().Analyze));
 
-        private static int Main(string[] args)
-        {
-            DebugUtil.WaitForDebuggerIfRequested(ref args);
-
-            try
-            {
-                return CommandLineApplication.Execute<Program>(args);
-            }
-            catch(PlatformNotSupportedException ex)
-            {
-                Console.Error.WriteLine(ex.Message);
-                return 1;
-            }
-            catch (OperationCanceledException)
-            {
-                return 0;
-            }
-        }
+        private static Argument DumpFile() =>
+            new Argument<FileInfo> { Name = "dumpFile", Description = "Name of the dump file to analyze." }.ExistingOnly();
+
+        private static Option RunCommand() =>
+            new Option(new[] { "-c", "--command" }, "Run the command on start.", new Argument<string[]>() { Name = "command" });
```

Comment:
There are a number of places where the implementation has extra or different functionality from what is in the spec. You should either propose spec updates to match implementation, or vice versa. All the folks looking at what we are building and providing feedback are commenting on the spec only - we can't expect to get much if any design feedback from code changes.

I'm OK checking work in progress that differs from the spec that is already here, but IMO the goal should be to bring them into agreement relatively quickly. In general I think the workflow should be that if you want to propose additional features your PR should start with a change to the spec. If the implementation is cheap enough I'm fine with a PR that modifies the spec and the implementation at the same time, but impl only design changes just makes it hard for us to get any feedback or community understanding of what is being built.

Reply:
If you need me to go through and comment on all the differences I can find I'm happy to do so, but I didn't attempt that right now.

Reply:
For `dotnet dump analyze`, it seems that the only difference is that I added the --command option. I can remove it from this PR or suggest adding it to your spec PR. I really haven't thought this option through anyway like should it still fall into interactive mode or should it be a set of initial commands that then falls into interactive.

If you are talking about the 5 interactive commands in you spec, then we should talk. I assumed that the set of commands would be the ones that "work" under the SOS host.

As far as `dotnet dump collect`, I did even attempt to match the spec yet, just get a basic command "template" that someone later could flesh out once the spec is done.

Reply:
I've commited the current spec PR, so now its actually reasonable for you to commit PRs to the spec whereas before that would have been understandably awkward.

The --command argument was one spot, but some others as I recall them: dump file naming format was different, set of SOS commands supported was different, extra commands in the interactive REPL such as listmodules, setthread, and exit. I'm not pushing back that any of these changes are bad, I just want to ensure they get full exposure in the spec.

Reply:
(Actually I might push back on the output naming format, but happy to chat about it)

Reply:
Should I change the spec with all the SOS commands support, the "native" commands like setthread, exit, etc now as part of my PR or later in a separate one?

Not sure what you mean "output naming format". The collect -o, --output option?

Reply:
I meant the name that you give to the dump file when the user didn't specify it. You've got a format string with the date and time in it, the spec has:

    -o, --output
        The path where collected dumps should be written. Defaults to .\dumpNNNN.dmp on windows and ./coreNNNN on
        Linux\Mac if not specified. NNNN is an increasing 4 digit counter for each dump, for example 
        .\dump0003.dmp. If the output_dump_path specifies a directory then dump files are written to that directory
        with the same dumpNNNN[.dmp] naming format. Specifying an exact filename is only permitted when capturing a
        single dump.

I'd suggest doing a PR to the spec separately because feedback can be slow and you probably don't want to hold up all the implementation work that is already here. For the SOS command part I think just a list of the pre-existing commands would be fine, but for the new commands you are adding more detail is probably merited.

Reply:
The dotnet dump collect options and code I really didn't change from the prototype.  I was focusing on dotnet dump analyze.  I'll write up an issue to make dump collect up to spec.